File: NSTU-lab3/lab3_go/containers/cuckoo.go
package containers

import (
	"fmt"
	"hash/fnv"
)

// CuckooHashMap - хеш-таблица с кукушкой хешированием
type CuckooHashMap struct {
	table1   []cuckooEntry
	table2   []cuckooEntry
	capacity int
	count    int
}

type cuckooEntry struct {
	key      string
	value    string
	occupied bool
}

const (
	cuckooDefaultCapacity = 101
	maxCuckooIterations   = 500
)

// NewCuckooHashMap создаёт новую хеш-таблицу с кукушкой
func NewCuckooHashMap(initialCapacity int) *CuckooHashMap {
	if initialCapacity <= 0 {
		initialCapacity = cuckooDefaultCapacity
	}
	return &CuckooHashMap{
		table1:   make([]cuckooEntry, initialCapacity),
		table2:   make([]cuckooEntry, initialCapacity),
		capacity: initialCapacity,
		count:    0,
	}
}

// hash1 первая хеш-функция
func (ch *CuckooHashMap) hash1(key string) int {
	hasher := fnv.New32a()
	hasher.Write([]byte(key))
	return int(hasher.Sum32()) % ch.capacity
}

// hash2 вторая хеш-функция
func (ch *CuckooHashMap) hash2(key string) int {
	hasher := fnv.New32a()
	hasher.Write([]byte(key))
	return int(hasher.Sum32()/uint32(ch.capacity)) % ch.capacity
}

// insertInternal вспомогательная функция для вставки с ограничением итераций
func (ch *CuckooHashMap) insertInternal(key string, value string, iterations int) error {
	if iterations >= maxCuckooIterations {
		return fmt.Errorf("слишком много итераций, требуется рехеширование")
	}

	idx1 := ch.hash1(key)

	// Пробуем table1
	if !ch.table1[idx1].occupied {
		ch.table1[idx1] = cuckooEntry{key: key, value: value, occupied: true}
		return nil
	}

	// Если ключ уже есть в table1, обновляем
	if ch.table1[idx1].key == key {
		ch.table1[idx1].value = value
		return nil
	}

	// Вытесняем из table1
	oldKey := ch.table1[idx1].key
	oldValue := ch.table1[idx1].value
	ch.table1[idx1] = cuckooEntry{key: key, value: value, occupied: true}

	idx2 := ch.hash2(oldKey)

	// Пробуем table2
	if !ch.table2[idx2].occupied {
		ch.table2[idx2] = cuckooEntry{key: oldKey, value: oldValue, occupied: true}
		return nil
	}

	// Если ключ уже есть в table2, обновляем
	if ch.table2[idx2].key == oldKey {
		ch.table2[idx2].value = oldValue
		return nil
	}

	// Вытесняем из table2
	nextKey := ch.table2[idx2].key
	nextValue := ch.table2[idx2].value
	ch.table2[idx2] = cuckooEntry{key: oldKey, value: oldValue, occupied: true}

	// Рекурсивно вставляем вытесненный элемент
	return ch.insertInternal(nextKey, nextValue, iterations+1)
}

// Put добавляет или обновляет пару ключ-значение
func (ch *CuckooHashMap) Put(key string, value string) error {
	if ch.count >= ch.capacity*2 {
		return fmt.Errorf("хеш-таблица заполнена")
	}

	// Проверяем, есть ли уже такой ключ
	if ch.Contains(key) {
		idx1 := ch.hash1(key)
		idx2 := ch.hash2(key)

		if ch.table1[idx1].occupied && ch.table1[idx1].key == key {
			ch.table1[idx1].value = value
			return nil
		}
		if ch.table2[idx2].occupied && ch.table2[idx2].key == key {
			ch.table2[idx2].value = value
			return nil
		}
	}

	err := ch.insertInternal(key, value, 0)
	if err == nil {
		ch.count++
	}
	return err
}

// Get возвращает значение по ключу
func (ch *CuckooHashMap) Get(key string) (string, error) {
	idx1 := ch.hash1(key)
	idx2 := ch.hash2(key)

	if ch.table1[idx1].occupied && ch.table1[idx1].key == key {
		return ch.table1[idx1].value, nil
	}
	if ch.table2[idx2].occupied && ch.table2[idx2].key == key {
		return ch.table2[idx2].value, nil
	}
	return "", fmt.Errorf("ключ не найден")
}

// Contains проверяет наличие ключа
func (ch *CuckooHashMap) Contains(key string) bool {
	idx1 := ch.hash1(key)
	idx2 := ch.hash2(key)

	return (ch.table1[idx1].occupied && ch.table1[idx1].key == key) ||
		(ch.table2[idx2].occupied && ch.table2[idx2].key == key)
}

// Remove удаляет пару по ключу
func (ch *CuckooHashMap) Remove(key string) bool {
	idx1 := ch.hash1(key)
	idx2 := ch.hash2(key)

	if ch.table1[idx1].occupied && ch.table1[idx1].key == key {
		ch.table1[idx1].occupied = false
		ch.count--
		return true
	}
	if ch.table2[idx2].occupied && ch.table2[idx2].key == key {
		ch.table2[idx2].occupied = false
		ch.count--
		return true
	}
	return false
}

// Clear очищает таблицу
func (ch *CuckooHashMap) Clear() {
	for i := range ch.table1 {
		ch.table1[i].occupied = false
	}
	for i := range ch.table2 {
		ch.table2[i].occupied = false
	}
	ch.count = 0
}

// Size возвращает количество элементов
func (ch *CuckooHashMap) Size() int {
	return ch.count
}

// Print выводит хеш-таблицу
func (ch *CuckooHashMap) Print() string {
	result := "CuckooHashMap {\n  Table 1:\n"
	for i, e := range ch.table1 {
		if e.occupied {
			result += fmt.Sprintf("    [%d] %s => %s\n", i, e.key, e.value)
		}
	}
	result += "  Table 2:\n"
	for i, e := range ch.table2 {
		if e.occupied {
			result += fmt.Sprintf("    [%d] %s => %s\n", i, e.key, e.value)
		}
	}
	result += fmt.Sprintf("} (size: %d)", ch.count)
	return result
}

File: NSTU-lab3/lab3/src/containers/cuckoo.cpp
#ifndef CUCKOO_CPP
#define CUCKOO_CPP

#include "hash.h"

template<typename K, typename V>
CuckooHashMap<K, V>::Entry::Entry() : occupied(false) {}

template<typename K, typename V>
CuckooHashMap<K, V>::Entry::Entry(const K& k, const V& v) : key(k), value(v), occupied(true) {}

template<typename K, typename V>
CuckooHashMap<K, V>::CuckooHashMap(size_t initialCapacity) 
    : capacity(initialCapacity), count(0) {
    table1.resize(capacity);
    table2.resize(capacity);
}

template<typename K, typename V>
size_t CuckooHashMap<K, V>::hash1(const K& key) const {
    std::hash<K> hasher;
    return hasher(key) % capacity;
}

template<typename K, typename V>
size_t CuckooHashMap<K, V>::hash2(const K& key) const {
    std::hash<K> hasher;
    return (hasher(key) / capacity) % capacity;
}

template<typename K, typename V>
bool CuckooHashMap<K, V>::insertInternal(const K& key, const V& value, int iterations) {
    if (iterations >= MAX_ITERATIONS) {
        throw std::runtime_error("CuckooHashMap: слишком много итераций, требуется рехеширование");
    }
    
    size_t idx1 = hash1(key);
    
    if (!table1[idx1].occupied) {
        table1[idx1] = Entry(key, value);
        return true;
    }
    
    if (table1[idx1].key == key) {
        table1[idx1].value = value;
        return true;
    }
    
    K oldKey = table1[idx1].key;
    V oldValue = table1[idx1].value;
    table1[idx1] = Entry(key, value);
    
    size_t idx2 = hash2(oldKey);
    
    if (!table2[idx2].occupied) {
        table2[idx2] = Entry(oldKey, oldValue);
        return true;
    }
    
    if (table2[idx2].key == oldKey) {
        table2[idx2].value = oldValue;
        return true;
    }
    
    K nextKey = table2[idx2].key;
    V nextValue = table2[idx2].value;
    table2[idx2] = Entry(oldKey, oldValue);
    
    return insertInternal(nextKey, nextValue, iterations + 1);
}

template<typename K, typename V>
void CuckooHashMap<K, V>::put(const K& key, const V& value) {
    if (count >= capacity * 2) {
        throw std::runtime_error("CuckooHashMap заполнена");
    }
    
    if (contains(key)) {
        size_t idx1 = hash1(key);
        size_t idx2 = hash2(key);
        
        if (table1[idx1].occupied && table1[idx1].key == key) {
            table1[idx1].value = value;
        } else if (table2[idx2].occupied && table2[idx2].key == key) {
            table2[idx2].value = value;
        }
        return;
    }
    
    if (insertInternal(key, value, 0)) {
        count++;
    }
}

template<typename K, typename V>
V CuckooHashMap<K, V>::get(const K& key) const {
    size_t idx1 = hash1(key);
    size_t idx2 = hash2(key);
    
    if (table1[idx1].occupied && table1[idx1].key == key) {
        return table1[idx1].value;
    }
    if (table2[idx2].occupied && table2[idx2].key == key) {
        return table2[idx2].value;
    }
    throw std::runtime_error("Ключ не найден");
}

template<typename K, typename V>
bool CuckooHashMap<K, V>::contains(const K& key) const {
    size_t idx1 = hash1(key);
    size_t idx2 = hash2(key);
    
    return (table1[idx1].occupied && table1[idx1].key == key) ||
           (table2[idx2].occupied && table2[idx2].key == key);
}

template<typename K, typename V>
bool CuckooHashMap<K, V>::remove(const K& key) {
    size_t idx1 = hash1(key);
    size_t idx2 = hash2(key);
    
    if (table1[idx1].occupied && table1[idx1].key == key) {
        table1[idx1].occupied = false;
        count--;
        return true;
    }
    if (table2[idx2].occupied && table2[idx2].key == key) {
        table2[idx2].occupied = false;
        count--;
        return true;
    }
    return false;
}

template<typename K, typename V>
void CuckooHashMap<K, V>::clear() {
    for (auto& entry : table1) {
        entry.occupied = false;
    }
    for (auto& entry : table2) {
        entry.occupied = false;
    }
    count = 0;
}

template<typename K, typename V>
size_t CuckooHashMap<K, V>::size() const {
    return count;
}

template<typename K, typename V>
bool CuckooHashMap<K, V>::empty() const {
    return count == 0;
}

template<typename K, typename V>
void CuckooHashMap<K, V>::print(std::ostream& os) const {
    os << "CuckooHashMap {\n  Table 1:\n";
    for (size_t i = 0; i < capacity; i++) {
        if (table1[i].occupied) {
            os << "    [" << i << "] " << table1[i].key << " => " << table1[i].value << "\n";
        }
    }
    os << "  Table 2:\n";
    for (size_t i = 0; i < capacity; i++) {
        if (table2[i].occupied) {
            os << "    [" << i << "] " << table2[i].key << " => " << table2[i].value << "\n";
        }
    }
    os << "} (size: " << count << ")";
}

#endif
